lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -482,10 +482,105 @@ struct WarpOpTransferWrite : public OpRewritePattern<vector::TransferWriteOp> {
   DistributionMapFn distributionMapFn;
 };
 
+/// Sink scf.for region out of WarpSingleOp. This can be done only if the
+/// scf.ForOp is the last operation in the region so that it doesn't change the
+/// order of execution. This creates a new scf.for region after the
+/// WarpSingleOp. The new scf.for region will contain a new WarpSingleLaneOp
+/// region. Example:
+/// ```
+/// %w = vector_ext.warp_execute_on_lane_0(%laneid) -> (vector<4xf32>) {
+///   ...
+///   %v1 = scf.for %arg3 = %c0 to %c128 step %c1 iter_args(%arg4 = %v)
+///   -> (vector<128xf32>) {
+///     ...
+///     scf.yield %r : vector<128xf32>
+///   }
+///   vector_ext.yield %v1 : vector<128xf32>
+/// }
+/// ```
+/// To:
+/// %w0 = vector_ext.warp_execute_on_lane_0(%arg0) -> (vector<4xf32>) {
+///   ...
+///   vector_ext.yield %v : vector<128xf32>
+/// }
+/// %w = scf.for %arg3 = %c0 to %c128 step %c1 iter_args(%varg = %q0)
+///   -> (vector<4xf32>) {
+///     %iw = vector_ext.warp_execute_on_lane_0(%laneid)
+///     args(%varg : vector<4xf32>) -> (vector<4xf32>) {
+///     ^bb0(%arg: vector<128xf32>):
+///       ...
+///       vector_ext.yield %ir : vector<128xf32>
+///     }
+///     scf.yield %iw : vector<4xf32>
+///  }
+/// ```
+struct WarpOpScfForOp : public OpRewritePattern<WarpSingleLaneOp> {
```

Comment:
This seems to do 2 things at the same time and feels like it would benefit from being split.
Step 1: splice the vector_ext.warp_execute_on_lane_0 just before the last for loop.
Step 2: swap the unique scf.for with the enclosing vector_ext.warp_execute_on_lane_0.

The first functionality in step 1 seems generally useful; don't we already have something like this?

Reply:
This is a great point. I had not seen it that way but it is much simpler indeed. The transformation of slicing the `vector_ext.warp_execute_on_lane_0` will be needed in more cases as you assumed. For instance this is also what is done in Matthias' PR: https://github.com/google/iree-llvm-sandbox/pull/303
What I'm wondering now is whether it should be a helper or a pattern on its own. Starting with a common helper is probably the way to go.
I'll push the code this way and then refactor this to create a helper that slice `vector_ext.warp_execute_on_lane_0`